keel-sql/src/main/resources/db/changelog/20210208-versioned-resources.yml

```diff
@@ -0,0 +1,47 @@
+databaseChangeLog:
+- changeSet:
+    id: versioned-resources
+    author: fletch
+    changes:
+    - addColumn:
+        tableName: resource
+        columns:
+        - name: version
+          type: integer
+          defaultValueNumeric: 1
+          afterColumn: id
+          constraints:
+            nullable: false
+    - dropUniqueConstraint:
+        tableName: resource
+        constraintName: name
+    - addUniqueConstraint:
+        tableName: resource
+        constraintName: resource_id_version_idx
+        columnNames: id, version
+    - dropView:
+        viewName: resource_with_metadata
+    - createView:
+        viewName: resource_with_metadata
+        selectQuery: |
+          select
+            resource.uid,
+            resource.id,
+            resource.application,
+            resource.kind,
+            json_object(
+              'uid', resource.uid,
+              'id', resource.id,
+              'version', resource.version,
+              'application', resource.application,
+              'environment', environment.uid,
+              'environmentName', environment.name,
+              'deliveryConfig', delivery_config.uid,
+              'serviceAccount', delivery_config.service_account
+            ) as metadata,
+            resource.spec
+          from resource
+          left outer join environment_resource on resource.uid = environment_resource.resource_uid
```

Comment:
Hmm, this implies that the relationship is between an environment and specific versions of a resource, which is not the case. While I don't like using `resource.id` as an FK, shouldn't that be what we use here instead of `resource.uid`?

Reply:
That _is_ the relationship. An environment doesn't contain all the versions of a resource, just the latest one.

Future work will define multiple versions of environments, so if a resource updates it will trigger the creation of a new environment version.

Reply:
Ah, got it. Makes sense. I was reading this thinking of the relationships as defined in the delivery config (where there's no concept of versions).